validator/db/manage.go

```diff
@@ -0,0 +1,137 @@
+)
+
+type epochProposals struct {
+	Epoch     []byte
+	Proposals []byte
+}
+
+type pubKeyProposals struct {
+	PubKey    []byte
+	Proposals []epochProposals
+}
+
+type pubKeyAttestations struct {
+	PubKey       []byte
+	Attestations []byte
+}
+
+// Merge merges data from sourceStores into a new store, which is created in targetDirectory.
+func Merge(ctx context.Context, sourceStores []*Store, targetDirectory string) error {
+	ctx, span := trace.StartSpan(ctx, "Validator.Db.Manage")
+	defer span.End()
+
+	var allProposals []pubKeyProposals
+	var allAttestations []pubKeyAttestations
+
+	for _, store := range sourceStores {
+		if err := store.db.View(func(tx *bolt.Tx) error {
+			proposalsBucket := tx.Bucket(historicProposalsBucket)
+			if err := proposalsBucket.ForEach(func(pubKey, _ []byte) error {
+				pubKeyProposals, err := getPubKeyProposals(pubKey, proposalsBucket)
+				if err != nil {
+					return errors.Wrapf(err,"Could not retrieve proposals for database in %s", store.databasePath)
+				}
+				allProposals = append(allProposals, *pubKeyProposals)
+				return nil
+			}); err != nil {
+				return errors.Wrapf(err, "Could not retrieve proposals for database in %s", store.databasePath)
+			}
+
+			attestationsBucket := tx.Bucket(historicAttestationsBucket)
+			if err := attestationsBucket.ForEach(func(pubKey, v []byte) error {
+				attestations := pubKeyAttestations{
+					PubKey:       make([]byte, len(pubKey)),
+					Attestations: make([]byte, len(v)),
+				}
+				copy(attestations.PubKey, pubKey)
+				copy(attestations.Attestations, v)
+				allAttestations = append(allAttestations, attestations)
+				return nil
+			}); err != nil {
+				return errors.Wrapf(err, "Could not retrieve attestations for database in %s", store.databasePath)
+			}
+
+			return nil
+		}); err != nil {
+			return err
+		}
+	}
+
+	if err := createTargetStore(targetDirectory, allProposals, allAttestations); err != nil {
+		return errors.Wrapf(err, "Could not create target store")
+	}
+
+	return nil
+}
+
+func getPubKeyProposals(pubKey []byte, proposalsBucket *bolt.Bucket) (*pubKeyProposals, error) {
+	pubKeyProposals := pubKeyProposals{
+		PubKey:    pubKey,
+		Proposals: []epochProposals{},
+	}
+
+	pubKeyBucket := proposalsBucket.Bucket(pubKey)
+	if err := pubKeyBucket.ForEach(func(epoch, v []byte) error {
+		epochProposals := epochProposals{
+			Epoch:     make([]byte, len(epoch)),
+			Proposals: make([]byte, len(v)),
+		}
+		copy(epochProposals.Epoch, epoch)
+		copy(epochProposals.Proposals, v)
+		pubKeyProposals.Proposals = append(pubKeyProposals.Proposals, epochProposals)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return &pubKeyProposals, nil
+}
+
+func createTargetStore(
+	targetDirectory string,
+	allProposals []pubKeyProposals,
+	allAttestations []pubKeyAttestations) error {
+
+	newStore, err := NewKVStore(targetDirectory)
+	defer func() {
+		if e := newStore.Close(); e != nil {
+			err = errors.Wrap(err, "Could not close the merged database")
```

Comment:
again this does nothing

Reply:
By "nothing" you mean that the errors are not handled, right? The code does close the db connections.

Reply:
Yeap, you are defining the errors but doing nothing with them. You should log an error here.